splitgpg2/__init__.py

```diff
@@ -0,0 +1,804 @@
+                    return
+            except FileNotFoundError:
+                pass
+
+        short_msg = "split-gpg2: '{}' wants to execute {}".format(
+            self.client_domain, name)
+        question = '{}\nDo you want to allow this{}?'.format(
+            short_msg,
+            'for the next {}s'.format(delay) if delay is not None else '')
+        if subprocess.call(['zenity', '--question', '--title', short_msg,
+                            '--text', question]) != 0:
+            raise Filtered
+
+        self.notify('command {} allowed'.format(name))
+        timestamp_path.touch()
+
+    def timestamp_path(self, name) -> pathlib.Path:
+        return pathlib.Path('{}_qubes-split-gpg2-timestamp_{}'.format(
+            self.agent_socket_path, name))
+
+    def client_write(self, data):
+        self.log_io('C <<<', data)
+        self.client_transport.write(data)
+
+    def eof_received(self) -> Optional[bool]:
+        # close connection to the real gpg agent too
+        if self.agent_transport is not None:
+            self.agent_transport.write_eof()
+        # automatically close the transport
+        return False
+
+    def data_received(self, untrusted_data: bytes) -> None:
+        # pylint: disable=arguments-differ
+        if len(self.untrusted_input_buffer) > ASSUAN_LINELENGTH*2:
+            self.log.error('Too much data received, dropping')
+
+        self.log_io('C >>>', untrusted_data)
+
+        self.untrusted_input_buffer.extend(untrusted_data)
+        if self.state == ServerState.agent_response:
+            # now is an agent turn, only buffer the data, but don't act on it
+            return
+
+        while b'\n' in self.untrusted_input_buffer:
+            # there is a full command buffered
+            untrusted_line, untrusted_rest = \
+                self.untrusted_input_buffer.split(b'\n', 1)
+            self.untrusted_input_buffer = untrusted_rest
+            # convert from bytesarray to bytes to be hashable
+            untrusted_line = bytes(untrusted_line)
+            if self.state == ServerState.client_command:
+                asyncio.ensure_future(self.handle_command(
+                    untrusted_line=untrusted_line))
+            elif self.state == ServerState.client_inquire:
+                asyncio.ensure_future(self.handle_inquire(
+                    untrusted_line=untrusted_line))
+
+    async def send_inquire(self, inquire, inquire_commands):
+        if self.state != ServerState.agent_response:
+            raise ProtocolError(
+                'invalid server state: INQUIRE not originating from gpg-agent')
+        loop = asyncio.get_event_loop()
+        self.inquire_complete = loop.create_future()
+        self.inquire_commands = inquire_commands
+        self.state = ServerState.client_inquire
+        self.client_write(b'INQUIRE ' + inquire + b'\n')
+        await self.inquire_complete
+        self.inquire_commands = {}
+        self.state = ServerState.agent_response
```

Comment:
Does this work? What if `inquires_complete` gets set but there's new data before we get to execute this code here? (Maybe I'm just confused by the async stuff)

Reply:
Good point. I'll move state change to handle_inquire and leave a comment here.